gui.md

```diff
@@ -0,0 +1,917 @@
+~~~lisp
+(with-ltk ()
+  (let ((button (make-instance 'button :text "hello")))
+    (grid button 0 0)))
+~~~
+
+That's all there is to it.
+
+
+### Reacting to events
+
+Many widgets have a `:command` argument that accept a lambda which is
+executed when the widget's event is started. In the case of a button,
+that will be on a click:
+
+~~~lisp
+(make-instance 'button
+  :text "Hello"
+  :command (lambda ()
+             (format t "clicked")))
+~~~
+
+
+### Interactive development
+
+When we start the Tk process in the background with `(start-wish)`, we
+can create widgets and place them on the grid interactively.
+
+See [the documentation](http://www.peter-herth.de/ltk/ltkdoc/node8.html).
+
+Once we're done, we can `(exit-wish)`.
+
+
+### Nodgui
+
+To try the Nodgui demo, do:
+
+~~~lisp
+(ql:quickload :nodgui)
+(nodgui.demo:demo)
+~~~
+
+## Qt4
+
+~~~lisp
+(ql:quickload '(:qtools :qtcore :qtgui))
+~~~
+
+~~~lisp
+(defpackage #:qtools-test
+  (:use #:cl+qt)
+  (:export #:main))
+(in-package :qtools-test)
+(in-readtable :qtools)
+~~~
+
+We create our main widget that will contain the rest:
+
+~~~lisp
+(define-widget main-window (QWidget)
+  ())
+~~~
+
+We create an input field and a button inside this main widget:
+
+~~~lisp
+(define-subwidget (main-window name) (q+:make-qlineedit main-window)
+  (setf (q+:placeholder-text name) "Your name please."))
+~~~
+
+~~~lisp
+(define-subwidget (main-window go-button) (q+:make-qpushbutton "Go!" main-window))
+~~~
+
+We stack them horizontally:
+
+~~~lisp
+(define-subwidget (main-window layout) (q+:make-qhboxlayout main-window)
+  (q+:add-widget layout name)
+  (q+:add-widget layout go-button))
+~~~
+
+and we show them:
+
+~~~lisp
+(with-main-window
+  (make-instance 'main-window))
+~~~
+
+![](/assets/gui/qtools-intro.png)
+
+That's cool, but we don't react to the click event yet.
+
+### Reacting to events
+
+Reacting to events in Qt happens through signals and slots. Slots are
+functions that receive signals, and signals are event carriers.
+
+We create a signal named `name-set` to throw when the button is clicked:
+
+~~~lisp
+(define-signal (main-window name-set) (string))
+~~~
+
+We want it to create a message window to display the text we entered
+in the input field. The `new-name` parameter is the value received by
+the signal, it is of type `string`.
+
+~~~lisp
+(define-slot (main-window name-set) ((new-name string))
+  (declare (connected main-window (name-set string)))
+  (q+:qmessagebox-information main-window "Greetings" (format NIL "Good day to you, ~a!" new-name)))
+~~~
+
+And now we make our button react to the `pressed` and `return-pressed`
+events. It sends the text value of the "name" input field.
+
+~~~lisp
+(define-slot (main-window go-button) ()
+  (declare (connected go-button (pressed)))
+  (declare (connected name (return-pressed)))
+  (signal! main-window (name-set string) (q+:text name)))
+~~~
+
+You can run again
+
+~~~lisp
+(with-main-window
+  (make-instance 'main-window))
+~~~
+
```

Comment:
You should mention that it is not necessary to create an extra signal every time. You could just as well pack the code from the name-set slot into the go-button slot. In fact, most of the time you won't create extra signals of your own.

Reply:
got it, thanks.